fdbclient/MutationLogReader.actor.h

```diff
@@ -0,0 +1,124 @@
+
+#include <deque>
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "flow/flow.h"
+#include "flow/ActorCollection.h"
+#include "flow/actorcompiler.h" // has to be last include
+
+namespace mutation_log_reader {
+
+struct RangeResultBlock {
+	RangeResult result;
+	Version firstVersion; // version of first record
+	Version lastVersion; // version of last record
+	uint8_t hash; // points back to the PipelinedReader
+	int prefixLen; // size of keyspace, uid, and hash prefix
+	int indexToRead; // index of first unconsumed record
+
+	Standalone<RangeResultRef> consume();
+
+	bool empty() { return indexToRead == result.size(); }
+
+	bool operator<(const RangeResultBlock& r) const {
+		// We want a min heap. The standard C++ priority queue is a max heap.
+		return firstVersion > r.firstVersion;
+	}
+};
+
+class PipelinedReader {
+public:
+	PipelinedReader(uint8_t h, Version bv, Version ev, unsigned pd, Key p)
+	  : readerLimit(pd), hash(h), prefix(StringRef(&hash, sizeof(uint8_t)).withPrefix(p)), beginVersion(bv),
+	    endVersion(ev), currentBeginVersion(bv), pipelineDepth(pd) {}
+
+	void startReading(Database cx);
+	Future<Void> getNext(Database cx);
+	ACTOR static Future<Void> getNext_impl(PipelinedReader* self, Database cx);
+
+	void release() { readerLimit.release(); }
+
+	PromiseStream<RangeResultBlock> reads;
+	FlowLock readerLimit;
+	uint8_t hash;
+	Key prefix; // "\xff\x02/alog/UID/hash/" for restore, or "\xff\x02/blog/UID/hash/" for backup
+
+	Future<Void> done() { return reader; }
+
+private:
+	Version beginVersion, endVersion, currentBeginVersion;
+	unsigned pipelineDepth;
+	Future<Void> reader;
+};
+
+} // namespace mutation_log_reader
+
+class MutationLogReader : public ReferenceCounted<MutationLogReader> {
+public:
+	MutationLogReader() : finished(256) {}
+	MutationLogReader(Database cx, Version bv, Version ev, Key uid, Key beginKey, unsigned pd)
+	  : beginVersion(bv), endVersion(ev), prefix(uid.withPrefix(beginKey)), pipelineDepth(pd), finished(0) {
+		pipelinedReaders.reserve(256);
+		if (pipelineDepth > 0) {
+			for (int h = 0; h < 256; ++h) {
+				pipelinedReaders.emplace_back(new mutation_log_reader::PipelinedReader(
+				    (uint8_t)h, beginVersion, endVersion, pipelineDepth, prefix));
+				pipelinedReaders[h]->startReading(cx);
+			}
+		}
+	}
+
+	ACTOR static Future<Reference<MutationLogReader>> Create(Database cx,
+	                                                         Version bv,
+	                                                         Version ev,
+	                                                         Key uid,
+	                                                         Key beginKey,
+	                                                         unsigned pd) {
+		state Reference<MutationLogReader> self(new MutationLogReader(cx, bv, ev, uid, beginKey, pd));
+		wait(self->initializePQ(self.getPtr()));
+		return self;
+	}
+
+	Future<Standalone<RangeResultRef>> getNext();
```

Comment:
What's the contract of this function? It seems the versions can be returned *out-of-order*:
Let's say hash value `a` has been consumed, then a new read of version range `[300, 400)` is issued to DB, which takes a long time.
In the mean time, another hash value `b` then is consumed, and next read of version range `[300, 350)` returned and consumed. Then next read `[350, 400)` is issued and returned.
Now we end up with `[350, 400)` for hash value `b` being returned before `[300, 350)` for hash value being returned.

Is this kind of out-of-order desirable? I guess not. It's probably worth describing the exact contract and how it is implemented.

Reply:
No it won't happen. The theory behind is, the versions falling into the same 1M range will have the same hash. In getNext(), we always cut when we reach the closest 1M of the RangeResultBlock (see the consume() function).

Reply:
I think MutationLogReader's definition needs a comment block explaining what it does and the contract.  Most importantly, it does deliver perfectly ordered mutations, which is a key requirement for mutation application and ideally would be the case for all mutation read use cases.

In the backup logging mode where logs are read from the KeySpace, log file contents are NOT in strict version order.  Instead, it is relied on that Restore will re-order the mutations as they are staged in the database before application.  This is why mutation log filtering on load during restore is only "partial", because some lexically adjacent KV pairs with different chunks of the same message block for a single version could be non-adjacent in the log file stream.  

It would be better for log file contents to be strictly ordered, which would enable perfect filtering during restore (which will already work with the existing code).  Additionally it will work well with a future plan I have for scaling mutation application up from 1 proxy to N proxies and bypass the log staging step completely.

Reply:
Thanks for explaining this. It's important do document the contract of delivering data in perfect orders!

Reply:
Added some comments.